internal/cmd/backup.go

```diff
@@ -337,37 +374,177 @@ func backupCreateCmdFunc(cmd *cobra.Command, args []string) (err error) {
-		return fmt.Errorf("error finalizing progress bar: %w", err)
+		if err := writeProgress(progressFile, relsResp); err != nil {
+			return err
+		}
 	}
 
+	totalTime := time.Since(relationshipReadStart)
+	_ = bar.Finish()
+
 	log.Info().
-		Uint("encoded", relsEncoded).
-		Uint("processed", relsProcessed).
-		Uint64("perSecond", perSec(uint64(relsProcessed), totalTime)).
+		Uint64("processed", relsProcessed).
+		Uint64("filtered", relsFilteredOut).
+		Uint64("throughput", perSec(relsProcessed, totalTime)).
 		Stringer("duration", totalTime).
 		Msg("finished backup")
 
+	backupCompleted = true
 	return nil
 }
 
+// encoderForNewBackup creates a new encoder for a new zed backup file. It returns the ZedToken at which the backup
+// must be taken.
+func encoderForNewBackup(cmd *cobra.Command, c client.Client, backupFile *os.File) (*backupformat.Encoder, *v1.ZedToken, error) {
+	prefixFilter := cobrautil.MustGetString(cmd, "prefix-filter")
+
+	schemaResp, err := c.ReadSchema(cmd.Context(), &v1.ReadSchemaRequest{})
+	if err != nil {
+		return nil, nil, fmt.Errorf("error reading schema: %w", addSizeErrInfo(err))
+	} else if schemaResp.ReadAt == nil {
+		return nil, nil, fmt.Errorf("`backup` is not supported on this version of SpiceDB")
+	}
+	schema := schemaResp.SchemaText
+
+	// Remove any invalid relations generated from old, backwards-incompat
+	// Serverless permission systems.
+	if cobrautil.MustGetBool(cmd, "rewrite-legacy") {
+		schema = rewriteLegacy(schema)
+	}
+
+	// Skip any definitions without the provided prefix
+
+	if prefixFilter != "" {
+		schema, err = filterSchemaDefs(schema, prefixFilter)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	zedToken := schemaResp.ReadAt
+
+	encoder, err := backupformat.NewEncoder(backupFile, schema, zedToken)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error creating backup file encoder: %w", err)
+	}
+
+	return encoder, zedToken, nil
+}
+
+func writeProgress(progressFile *os.File, relsResp *v1.BulkExportRelationshipsResponse) error {
+	err := progressFile.Truncate(0)
+	if err != nil {
+		return fmt.Errorf("unable to truncate backup progress file: %w", err)
+	}
+
+	_, err = progressFile.Seek(0, 0)
+	if err != nil {
+		return fmt.Errorf("unable to seek backup progress file: %w", err)
+	}
+
+	_, err = progressFile.WriteString(relsResp.AfterResultCursor.Token)
+	if err != nil {
+		return fmt.Errorf("unable to write result cursor to backup progress file: %w", err)
+	}
+
+	return nil
+}
+
+// openProgressFile returns the progress marker file and the stored progress cursor if it exists, or creates
+// a new one if it does not exist. If the backup file exists, but the progress marker does not, it will return an error.
+//
+// The progress marker file keeps track of the last successful cursor received from the server, and is used to resume
+// backups in case of failure.
+func openProgressFile(backupFileName string, backupAlreadyExisted bool) (*os.File, *v1.Cursor, error) {
```

Comment:
could we somehow store this in the parquet file itself?

Reply:
It's not a parquet file, it's an Avro OCF file, and no: I clarify in the code that OCF are not meant for in-place updates, but for streaming append. I explored it.